tests-arquillian/src/test/java/org/jboss/weld/tests/contexts/propagation/ContextPropagationService.java

```diff
@@ -0,0 +1,111 @@
+        WeldManager get = CDI.current().select(WeldManager.class).get();
+        for (WeldAlterableContext context : CDI.current().select(WeldManager.class).get().getActiveWeldAlterableContexts()) {
+            // TODO shouldnt context.getAllContextualInstances() return Collection<ContextualInstance<Object>>> instead? what is T for there
+            scopeToContextualInstances.put(context.getScope(), context.getAllContextualInstances());
+        }
+        // We create a task wrapper which will make sure we have contexts propagated
+        Callable<T> wrappedTask = new Callable<T>() {
+
+            @Override
+            public T call() throws Exception {
+                WeldManager weldManager = CDI.current().select(WeldManager.class).get();
+                BoundRequestContext requestContext = weldManager.instance().select(BoundRequestContext.class, BoundLiteral.INSTANCE).get();
+                BoundSessionContext sessionContext = weldManager.instance().select(BoundSessionContext.class, BoundLiteral.INSTANCE).get();
+                BoundConversationContext conversationContext = weldManager.instance().select(BoundConversationContext.class, BoundLiteral.INSTANCE).get();
+
+                // we will be using bound contexts, prepare backing structures for contexts
+                Map<String, Object> sessionMap = new HashMap<>();
+                Map<String, Object> requestMap = new HashMap<>();
+                BoundRequest boundRequest = new MutableBoundRequest(requestMap, sessionMap);
+
+                // activate contexts
+                // TODO - for now let's just activate all the scopes we know should be there
+                // TODO - how to activate custom contexts?
+                requestContext.associate(requestMap);
+                requestContext.activate();
+                sessionContext.associate(sessionMap);
+                sessionContext.activate();
+                conversationContext.associate(boundRequest);
+                conversationContext.activate();
+
+                // propagate all contexts that have some bean in them
+                if (scopeToContextualInstances.get(requestContext.getScope()) != null) {
+                    requestContext.clearAndSet(scopeToContextualInstances.get(requestContext.getScope()));
+                }
+                if (scopeToContextualInstances.get(sessionContext.getScope()) != null) {
+                    sessionContext.clearAndSet(scopeToContextualInstances.get(sessionContext.getScope()));
+                }
+                if (scopeToContextualInstances.get(conversationContext.getScope()) != null) {
+                    conversationContext.clearAndSet(scopeToContextualInstances.get(conversationContext.getScope()));
+                }
+
+                // now execute the actual original task
+                T result = task.call();
+
+                // cleanup, context deactivation
+                // TODO - this can mess up state in original thread?
```

Comment:
I think this will trigger `@PreDestroy` calls (if there are any) on the beans. However, those beans may still be active in the original thread which would corrupt their state. Untested ATM.

Reply:
It might be better to just `deactivate()` the context, skipping the pre destroy calls here.
Should be cleared at MP conc spec level.

Reply:
I have decided to leave out `invalidate()` in this test with a comment explaining it. We can always chose otherwise but since this test is for now the only "how-to reference" for context propagation, I would like to have it minimal and clear.